### ISSUE TITLE
[FEATURE] Get caller instance object of event Raisers 

### ISSUE THREAD

**Issue by TheXRMonk** [unity-atoms/unity-atoms#376]
**Is your feature request related to a problem? Please describe.**
When debugging only the stacktrace is available - from which you can get no info on which instance of the type raised an event.

**Describe the solution you'd like**
I assume (from limited research) it's impossible to get the instance of the caller, but I would really like to be able to click a button in the event stacktrace debugger to highlight the instance object that made the raise. Like when you click the log message in the console to get pointed to the context.

**Describe alternatives you've considered**
Making my own AtomAction that I forcefully put on every gameobject that Raises any event, just to log a console message with the context to the GameObject.

**Comment by soraphis** [comment]
We can't do it via reflection (there is no way to find the calling object).

So only way would be to **change the API** for the [AtomEvent.cs ](https://github.com/unity-atoms/unity-atoms/blob/canary/Packages/Core/Runtime/Events/AtomEvents.cs) Raise method to accept a sender as an argument.

this needs to be passed through to the [StackTraceEntry ](https://github.com/unity-atoms/unity-atoms/blob/canary/Packages/Core/Runtime/StackTrace/StackTraceEntry.cs) into a new member field "_sender". 

In the [StackTraceEditor](https://github.com/unity-atoms/unity-atoms/blob/master/Packages/Core/Editor/Editors/StackTraceEditor.cs#L161-L163) as additional callback action the object can be [pinged](https://docs.unity3d.com/ScriptReference/EditorGUIUtility.PingObject.html).

**Comment by TheXRMonk** [comment]
> Since it most likely will be a breaking change, this would need to wait for the next major release though. Alternatively we **could** overload the Raise method.

Wouldn't this break the ability to raise events on unity events in the inspector? (Can't render more than one method parameter)?

**Comment by TheXRMonk** [comment]
An alternative would be to use the new Dependency viewer to find all used references of an Atom as a starting point for debugging issues;
![image](https://user-images.githubusercontent.com/54309738/204532917-dde00836-d292-4bec-8dfc-6105c42664f1.png)
However, the package is still experimental and just renaming an Atom breaks the dependency tree, forcing a rebuild, which then doesn't always captures all references within scenes. And even if this worked, I still wouldn't know on a per call basis which object instance Raised the event.

**Comment by TheXRMonk** [comment]
Alrighty, this seems to be the solution I'm going for for now;

```
public UnityEvent RaiseAtomEvent;

public void RaiseAtom()
{
    if (RaiseAtomEvent.GetPersistentEventCount() == 0)
    {
        Debug.LogError("No Atom assigned to be raised", this);
        return;
    }

    var targetObject = RaiseAtomEvent.GetPersistentTarget(0);

    if (targetObject is not AtomEventBase atomObject)
    {
        Debug.LogError("Only AtomEvents are allowed to be raised by RaiseAtomEvent!", this);
        return;
    }

    string atomType = atomObject.GetType().ToString().Split(".").Last();
    
    Debug.Log("[ATOM] Raising " + atomType + " : " + atomObject.name + " on " + gameObject, gameObject);
    
    if (Application.isPlaying)
        RaiseAtomEvent.Invoke();
}
```
Along with;

```
public static void RaiseWithDebug<T>(this AtomEvent<T> atomObject, T value, Object context)
{
    string atomType = atomObject.GetType().ToString().Split(".").Last();
    
    Debug.Log("[ATOM] Raising " + atomType + " : " + atomObject.name + " on " + context.name, context);
    
    if (Application.isPlaying)
        atomObject.Raise(value);
}

```

And then just forcing the practice of never calling Raise() directly on an Atom.
Sucks, but works.

**Comment by soraphis** [comment]
> Wouldn't this break the ability to raise AtomEvents (with parameters) from unity events in the inspector? 

just pass null, or the atom itself.

**Comment by TheXRMonk** [comment]
> > Wouldn't this break the ability to raise AtomEvents (with parameters) from unity events in the inspector?
> 
> just pass null, or the atom itself.

Yes that's my point. You can't raise an atom with a value AND pass the caller using UnityEvents in the inspector. Which is one lf the main selling points of the whole architecture.

**Comment by soraphis** [comment]
> Which is one lf the main selling points of the **whole architecture.**

I would fight that statement any day!

> You can't raise an atom with a value AND pass the caller using UnityEvents in the inspector

Why wouldn't that work? It's a custom Inspector like any other. so if the Raise method is modified to receive a second parameter the inspector will be modified obviously too.

E.g. here:

https://github.com/unity-atoms/unity-atoms/blob/master/Packages/Core/Editor/Editors/Events/AtomEventEditor.cs#L28

just adding either null or `atomEvent` as a second parameter.

___________

Please note that everything I wrote is meant as notes for a future Pull Request, not as a solution you could just implement in on the consumer end of the library. (I thought that was clear, since the issue has the label `enhancement`)

**Comment by TheXRMonk** [comment]
>Why wouldn't that work? It's a custom Inspector like any other. so if the Raise method is modified to receive a second parameter the inspector will be modified obviously too.

Unity Events arent custom Inspectors. 
Take a button as an example. Add a listener to that in the inspector with a reference to your atomEvent and call Raise() om that. 
UnityEvents c

**Comment by TheXRMonk** [comment]
Sry for close - fat phone thumbs...

*opens git issue requesting a feature to revert "close with comment" entries*

**Comment by soraphis** [comment]
1. You can use Unity Events. 

![image](https://user-images.githubusercontent.com/3066582/204666259-a4cba236-ba72-413b-89f7-1c5eecbb9f6e.png)

See https://docs.unity3d.com/ScriptReference/Events.UnityEvent_2.html 
the arguments must be passed from code, though.

______
If you insinst on the inspector functionality, you can use something like https://openupm.com/packages/com.solidalloy.extevents/
______

A part of unity atoms - and the scriptable object architecture as outlined by Ryan Hipple - is to NOT rely on UnityEvent, that's the whole point of AtomEvents. ( For reference this part of the talk: https://youtu.be/raQ3iHhE_Kk?t=1764 )

So why not just create a 

```
[System.Serializable]
public struct StringEventWithArgs{
    public StringEventReference unityAtomsEvent;
    public string argument;
}
```

then you can modify the argument in the inspector as you wanted and you don't need to go through the UnityEvent to invoke a UnityAtomsEvent.

**Comment by TheXRMonk** [comment]
Wow we are going in circles. 

He in the video you linked litterally goes on to call the gameevent scriptable object from a unityevent like I've explained all along;
![image](https://user-images.githubusercontent.com/54309738/204730189-b1a952c4-a006-4d05-a2ca-c50c568a4a83.png)
@35:50

And you litterally just linked an alternative to my previous post (a replacement package to UnityEvent

**Comment by lgarczyn** [comment]
What about something like this

```


public class UnityEventDebug<T>: UnityEvent<T> {

    #if UNITY_EDITOR

    Object _parent = null;

    public void SetParent(Object parent) {
        _parent = parent;
    } 

    new void Invoke(T value) {
        
        for (int i = 0; i < GetPersistentEventCount(); i++) {
            string name = GetPersistentMethodName(i);
            Object target = GetPersistentTarget(i);
            if (typeof(UnityAtoms.IRaisable<T>).IsAssignableFrom(target.GetType())) {
                // call but with parent;
            } else {
                // call normally
            }
        }
    }
    #else
    public SetParent(Object parent) {}
    #endif
}
```

**Comment by TheOnlyMunk** [comment]
> What about something like this
> 
> ```cs
> public class UnityEventDebug<T>: UnityEvent<T> {
> 
>     #if UNITY_EDITOR
> 
>     Object _parent = null;
> 
>     public void SetParent(Object parent) {
>         _parent = parent;
>     } 
> 
>     new void Invoke(T value) {
>         
>         for (int i = 0; i < GetPersistentEventCount(); i++) {
>             string name = GetPersistentMethodName(i);
>             Object target = GetPersistentTarget(i);
>             if (typeof(UnityAtoms.IRaisable<T>).IsAssignableFrom(target.GetType()) && name == "RaiseFnName") {
>                 // call but with parent;
>             } else {
>                 // call normally
>             }
>         }
>     }
>     #else
>     public SetParent(Object parent) {}
>     #endif
> }
> ```

Appreciate the thought, but again, this wouldn't work with Unity's built-in components, unless I'm unaware of a way to inject this new type into components like the default button. 

Maybe some way similar to how https://github.com/handzlikchris/Unity.TransformChangesDebugger.API works, but I have no idea how that works.

**Comment by lgarczyn** [comment]
You could try assembly patching, but it's a pile of worm I think few people would get into.

**Comment by TheOnlyMunk** [comment]
Yea, seems like Harmony is the way he did it. Guess I would have to patch every unityevent<T>.

**Comment by soraphis** [comment]
Closing this one, as it is not really feasible. 
If someone stumbles over this and really has this debugging need, a workaround might be to assign the event-raiser first to a "static global" variable and in the event-debug log read from this static global variable.

But I don't see a way for this approach (while only working on the mainthread, but that would be an okay tradeoff) being built into the framework without a callers injection....

not without changing the API... 
or with relying on Code weaving (e.g. using Fody)
